### ISSUE TITLE
Detect concurrent backup runs

### ISSUE THREAD

**Issue by fd0** [restic/restic#711]
In https://github.com/restic/restic/issues/519#issuecomment-269347162 we discussed what happens when restic is run via cron, and a second instance is started before the first one terminates. This issue tracks adding code which detects this situation.

**Comment by mholt** [comment]
Would it be sufficient to just write some sort of file in the repo while a backup is running? Then just check to see if that file exists before starting? The chances of starting restic within _moments_ of each other such that there's a race condition for the file is highly unlikely (if you're worried about it, you could check for existence when starting and again just before you write the file out). Much more likely is the chance that restic is running a backup already.

In another thread it was suggested to make a configurable time out (before the next cron runs), but frankly I would never want a running backup to terminate just because we hope it will start up again soon.

**Comment by fd0** [comment]
We're already creating a lock file in the repo for all operations, you've run into this in #708. A lock file contains some information about the process (PID etc.): https://github.com/restic/restic/blob/master/src/restic/lock.go#L27-L33

When we add the list of directories (or even the complete command line) there it sounds doable.

**Comment by mlissner** [comment]
I'm surprised this issue is still here. I am setting up backups for a system with about 4TB of data, and this was one of the first concerns I had as I migrate our backups to restic. I can handle this in my bash script, but no way will I do it as well as a properly built in feature. For us, on a live server, it's a big deal if we get more and more concurrent backups running.

**Comment by fd0** [comment]
@mlissner you could also just use the standard utility `flock` for this task, e.g.:

```
$ flock --nonblock --conflict-exit-code 0 /run/user/$UID/restic.lock restic backup [...]
```

It'll start restic once, and refuses to start a second instance until the other one has finished. With `--nonblock` and `--conflict-exit-code 0` it will just silently return if the old instance is still running.

It does not feel to me as of such high priority as it seems to be for you, especially since there is a solution (using `flock`), but maybe I don't see the whole picture.

Does that work for you?

**Comment by mlissner** [comment]
Yeah, I ended up researching and using the flock recipe here: https://stackoverflow.com/a/169969/64911. 

But I guess what makes this a high priority for me is that it's the one thing that prevents resit from being plug and play. Like, the whole system is so elegant and simple, and it's more or less two steps:

1. Init
2. Backup
3. (set up cron job)

But when you get to the third step, there's a major undocumented gotcha. Maybe a good solution here is to add an FAQ so that people at least will realize there's a gotcha? Maybe it's worth adding a whole section on "scheduled backups" or something like that, which covers best practices for servers and cron jobs and things like this? It could also include things like ionice, and nice?

**Comment by fd0** [comment]
Sounds good, would you be interested to write some text for the manual? :)

**Comment by darkdragon-001** [comment]
The following situation should be thought about:
- The first backup run is stale for some reason (running but not advancing -> will run forever)
- All consecutive runs will not run because another one is running

This situation will basically disable backups without the user noticing!

**Comment by rawtaz** [comment]
My thoughts:

- What @darkdragon-001 wrote right before this comment should be detected by monitoring of the backups, so it shouldn't be a problem (in other words, anyone and everyone should monitor their backups and if they don't run detect that this is the case).

- Regarding the initial issue, that restic should detect if an existing backup run is active, I would rather see that restic *does* start a second backup run even if this is the case. Simply because if there's an existing run that hasn't finished for whatever reason, my chances of having restorable snapshots *increase* if I *do* start a second backup run in this case. Restic waiting for the first one to finish *decreases* the chances. Restic can't possibly decide what is a stalled backup run and not.

In summary I think it's better to just let restic start and run a backup when it's asked to, regardless of whether another one is already ongoing. If for some specific reasons/use cases one does not want that, then one should handle that oneself, e.g. using the `flock` method. This should be the safest approach.

**Comment by kekkoudesu** [comment]
> * Regarding the initial issue, that restic should detect if an existing backup run is active, I would rather see that restic _does_ start a second backup run even if this is the case. Simply because if there's an existing run that hasn't finished for whatever reason, my chances of having restorable snapshots _increase_ if I _do_ start a second backup run in this case. Restic waiting for the first one to finish _decreases_ the chances. Restic can't possibly decide what is a stalled backup run and not.
> 
> 
> In summary I think it's better to just let restic start and run a backup when it's asked to, regardless of whether another one is already ongoing. If for some specific reasons/use cases one does not want that, then one should handle that oneself, e.g. using the `flock` method. This should be the safest approach.

I've landed here after reading [this part of the docs](https://restic.readthedocs.io/en/stable/040_backup.html#scheduling-backups):

> When scheduling restic to run recurringly, please make sure to detect already running instances before starting the backup.

Did anything change in the past four years (I see this change was made in 2021)? This seems to contradict what you're saying here; as if running another instance of Restic will cause issues with the backup.

I have a bash script that does the following when run via `cron`:

* Checks if it has an exclusive lock on a file with `flock`.
* If so, runs `restic unlock` if for some reason the repository has been locked by Restic.
* Performs a `backup`, `check`s the data, and `prune`s/`forget`s backups according to a retention policy.

From what you're saying, it might be worth removing the first step with `flock`.

Am I doing something particularly dumb here? My inclusion of `restic unlock` comes from this script included on the Arch Wiki: https://wiki.archlinux.org/title/Restic#Backup_script

Is `restic unlock` only something that should be run manually and never in a script?

**Comment by MichaelEischer** [comment]
> When scheduling restic to run recurringly, please make sure to detect already running instances before starting the backup.

The goal here is to make sure that there are not too many `backup` instances that run in parallel. Each one consumes resources such that continuously starting new backups that become stuck again can eventually exhaust all available resources on the host. The repository format does not care if multiple backups happen in parallel.

> Is `restic unlock` only something that should be run manually and never in a script?

It should be safe to use since restic 0.10.0 (or so). In combination with `flock` it's guaranteed anyways that only a single instance of restic runs at a time on the host. That should already be enough to prevent any theoretically possible problems with `unlock`.

> In summary I think it's better to just let restic start and run a backup when it's asked to, regardless of whether another one is already ongoing.

The backup script should always be combined with some monitoring that the backup has recently completed successfully. Either by checking that a recent snapshot exists, or by tracking when `backup` has successfully (exit code 0 !!!) completed.

**Comment by kekkoudesu** [comment]
Thank you for the clarification! It's very reassuring that multiple `restic backup` instances won't clobber each other.

> The backup script should always be combined with some monitoring that the backup has recently completed successfully. Either by checking that a recent snapshot exists, or by tracking when backup has successfully (exit code 0 !!!) completed.

My script stores `$?` in a variable and depending on the exit status (based on [the documentation here](https://restic.readthedocs.io/en/stable/040_backup.html#exit-status-codes)) sends me a message after every backup, as the docs suggest:

> One can use these exit status codes in scripts and other automation tools, to make them aware of the outcome of the backup run. To manually inspect the exit code in e.g. Linux, run `echo $?`.
